src/pyhf/constraints.py

```diff
@@ -38,85 +44,105 @@ def _precompute(self):
             except AttributeError:
                 normal_constraint_sigmas.append([1.0] * len(thisauxdata))
 
-            normal_constraint_data.append(thisauxdata)
-            normal_constraint_mean_indices.append(self.par_indices[parslice])
-
-        if normal_constraint_mean_indices:
-            normal_mean_idc = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_mean_indices,
-                    )
-                )
+        if self.parameter_helper.index_selection:
+            self._normal_data = default_backend.astensor(
+                default_backend.concatenate(normal_constraint_data), dtype='int'
             )
-            normal_sigmas = default_backend.concatenate(
-                list(map(default_backend.astensor, normal_constraint_sigmas))
+
+            _normal_sigmas = default_backend.concatenate(normal_constraint_sigmas)
+            sigmas = default_backend.reshape(_normal_sigmas, (1, -1))  # (1, normals)
+            self._batched_sigmas = default_backend.tile(
+                sigmas, (self.batch_size or 1, 1)
             )
-            normal_data = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_data,
-                    )
+
+            # self.parameter_helper.index_selection  is [ (batch, parslice) ] ]
+            access_field = None
```

Comment:
this code should be refactored a bit more. This feels like something that could be done with a simply concatenate call against `axis=0` or something instead of an actual loop.

Reply:
good catch.. replace with concatenation of `axis=1`